internal/build/job/manager.go

```diff
@@ -47,19 +57,85 @@ func (jbm *jobManager) GarbageCollect(ctx context.Context, mod kmmv1beta1.Module
 	return deleteNames, nil
 }
 
-func (jbm *jobManager) Sync(ctx context.Context, mod kmmv1beta1.Module, m kmmv1beta1.KernelMapping, targetKernel string, targetImage string, pushImage bool) (build.Result, error) {
+func (jbm *jobManager) ShouldSync(
+	ctx context.Context,
+	mod kmmv1beta1.Module,
+	m kmmv1beta1.KernelMapping,
+	opts ...build.SyncOption) (bool, error) {
+
+	options := build.SyncOptions{}
+	for _, opt := range opts {
+		opt(&options)
+	}
+	if options.Owner == nil {
+		options.Owner = &mod
+	}
+
+	// if there is no build specified skip
+	if !module.ShouldBeBuilt(mod.Spec, m) {
+		return false, nil
+	}
+
+	targetImage := m.ContainerImage
+
+	// if build AND sign are specified, then we will build an intermediate image
+	// and let sign produce the one specified in targetImage
+	if module.ShouldBeSigned(mod.Spec, m) {
+		targetImage = module.IntermediateImageName(options.Owner.GetName(), options.Owner.GetNamespace(), targetImage)
+	}
+
+	// build is specified and targetImage is either the final image or the intermediate image
+	// tag, depending on whether sign is specified or not. Either way, if targetImage exists
+	// we can skip building it
+	exists, err := module.ImageExists(ctx, jbm.client, jbm.registry, mod.Spec, options.Owner.GetNamespace(), m, targetImage)
+	if err != nil {
+		return false, fmt.Errorf("failed to check existence of image %s: %w", targetImage, err)
+	}
+
+	return !exists, nil
+}
+
+func (jbm *jobManager) Sync(
+	ctx context.Context,
+	mod kmmv1beta1.Module,
+	m kmmv1beta1.KernelMapping,
+	targetKernel string,
+	opts ...build.SyncOption) (build.Result, error) {
+
+	options := build.SyncOptions{}
+	for _, opt := range opts {
+		opt(&options)
+	}
+	if options.Owner == nil {
+		options.Owner = &mod
+	}
+
 	logger := log.FromContext(ctx)
 
 	logger.Info("Building in-cluster")
 
-	buildConfig := jbm.helper.GetRelevantBuild(mod, m)
+	buildConfig := jbm.helper.GetRelevantBuild(mod.Spec, m)
```

Comment:
WDYT about moving all the decision for buildConfig and job Option into the MakeJobTemplate function of the maker?  Since that function create and Job and makes all the decision regarding how it should be set, maybe it needs to take those decisions also. this way the Sync code will be more clear and readable and will deal only with managing the job